docs/DafnyRef/UserGuide.md

```diff
@@ -315,7 +315,260 @@ This list is not exhaustive but can definitely be useful to provide the next ste
+   assert X;
+   if b {
+     assert Y;
+   } else {
+     assert Z;
+     assert P;
+   }
+   assert W;
+}
+```
+
+This verifies instantly. This gives use a strategy to bissect, or do binary search to find the assertion that slows everything down.
+Now, we move the `assume false;` below the next assertion:
+
+```dafny
+method NotTerminating() {
+   assert X;
+   assume false;
+   if b {
+     assert Y;
+   } else {
+     assert Z;
+     assert P;
+   }
+   assert W;
+}
+```
+
+If verification is slow again, we can use [techniques seen before](#sec-verification-debugging) to decompose the assertion and find which component is slow to prove.
+
+If verification is fast, that's the sign that `X` is probably not the problem,. We now move the `assume false;` after the if/then block:
+
+```dafny
+method NotTerminating() {
+   assert X;
+   if b {
+     assert Y;
+   } else {
+     assert Z;
+     assert P;
+   }
+   assume false;
+   assert W;
+}
+```
+
+Now, if verification is fast, we know that `assert W;` is the problem. If it is slow, we know that one of the two branches of the `if` is the problem.
+The next step is to put an `assume false;` at the end of the `then` branch, and an `assume false` at the beginning of the else branch:
+
+```dafny
+method NotTerminating() {
+   assert X;
+   if b {
+     assert Y;
+     assume false;
+   } else {
+     assume false;
+     assert Z;
+     assert P;
+   }
+   assert W;
+}
+```
+
+Now, if verification is slow, it means that `assert Y;` is the problem.
+If verification is fast, it means that the problem lies in the `else` branch.
+One trick to ensure we measure the verification time of the `else` branch and not the then branch is to move the first `assume false;` to the top of the then branch, along with a comment indicating that we are short-circuiting it for now.
+Then, we can move the second `assume false;` down and identify which of the two assertions makes the verifier to be slow.
+
+
+```dafny
+method NotTerminating() {
+   assert X;
+   if b {
+     assume false; // Short-circuit because this branch is verified anyway
+     assert Y;
+   } else {
+     assert Z;
+     assume false;
+     assert P;
+   }
+   assert W;
+}
+```
+
+If verification is fast, which of the two assertions `assert Z;` or `assert P;` causes the slowdown?[^answer-slowdown]
+
+[^answer-slowdown]: `assert P;`.
+
+We now hope you know enough of `assume false;` to locate an assertion that makes verification slow.
+Nest, we will describe some other strategies at the assertion level to figure out what happens and perhaps fix it.
+
+#### 24.8.2.2. `assert ... by {}` {#sec-verification-debugging-assert-by}
+
+If an assertion `assert X;` is slow, it is possible that calling a lemma or invoking other assertions can help proving it: The postcondition of this lemma, or the added assertions, could help the Dafny verifier figure out faster how to prove the result.
+
+```dafny
+  assert LEMMA_PRECONDITION;
```

Comment:
This assert is unnecessary in most cases, since it's exactly the precondition of the following function.

Reply:
```dafny
assert SOMETHING_HELPING_TO_PROVE_LEMMA_PRECONDITION;
```